feat: implement stateless Wildmeta UserOp authentication

silva-fj opened the pull request:
## Summary
This PR implements a new RPC method `omni_submitUserOpWithAuth` that enables stateless client authentication for Wildmeta clients. It replaces JWT-based authentication with cryptographic signature verification linked to Hyperliquid agent wallets.

## Key Changes

### 1. New Authentication Method (`ClientAuth::WildmetaHl`)
- Added new variant to support Hyperliquid-linked authentication
- Includes agent address, main address, signature, and business JSON payload

### 2. Wildmeta API Integration
- Created `wildmeta-api` crate with HTTP client implementation
- Implements `verify_hyperliquid_link` endpoint to check if addresses are linked
- Configurable via `OE_WILDMETA_API_URL` environment variable
- Falls back to mock implementation when URL not configured

### 3. RPC Method Implementation (`omni_submitUserOpWithAuth`)
- Verifies cryptographic signatures from agent wallets
- Validates Hyperliquid link between agent and main addresses
- Includes timestamp-based replay protection
- Maintains compatibility with existing UserOp validation logic

### 4. Storage Layer
- Added `WildmetaTimestampStorage` for monotonic timestamp verification
- Prevents replay attacks by tracking last timestamp per main address

## Configuration
Set `OE_WILDMETA_API_URL` environment variable to enable real API integration:
```bash
export OE_WILDMETA_API_URL="https://api.wildmeta.example.com"
```

kziemianek committed:
omni_submitUserOp rpc method

kziemianek committed:
accept multiple userops at once

kziemianek committed:
entry point client wallet as beneficiary in aa

kziemianek committed:
omni_implementation_address -> omni_wallet_implementation_address

kziemianek committed:
OmniSubmitUserOp -> SubmitUserOp

kziemianek committed:
remove redundant Chain enum

kziemianek committed:
remove Chain name() function

kziemianek committed:
provide entry point contract address

kziemianek committed:
remove userop hashes from response

kziemianek committed:
remove omni_clients_map

kziemianek committed:
improve UserOp validation

kziemianek committed:
remove unused variables

kziemianek committed:
return tx

kziemianek committed:
userop simulations

kziemianek committed:
Merge branch 'dev' into p-1602-worker-code-to-initiate-the-deposit

kziemianek committed:
Merge branch 'dev' into p-1602-worker-code-to-initiate-the-deposit

kziemianek committed:
remove unnecessary code

kziemianek committed:
accept wallet_index

kziemianek committed:
bytecode caching

kziemianek committed:
Merge branch 'dev' into p-1602-worker-code-to-initiate-the-deposit

kziemianek committed:
Merge branch 'dev' into p-1602-worker-code-to-initiate-the-deposit

kziemianek committed:
Merge branch 'dev' into p-1602-worker-code-to-initiate-the-deposit

kziemianek committed:
omni cli

kziemianek committed:
Merge branch 'dev' into p-1602-worker-code-to-initiate-the-deposit

kziemianek committed:
remove session related fields from UserOp struct

kziemianek committed:
adjust UserOp json representation

kziemianek committed:
make signature optional

kziemianek committed:
replace Chain enum with u64

kziemianek committed:
Merge branch 'dev' into p-1602-worker-code-to-initiate-the-deposit

kziemianek committed:
fix clippy

silva-fj committed:
Add WildmetaHl variant to ClientAuth enum

silva-fj committed:
Add WildmetaApi trait and mock implementation

silva-fj committed:
Add WildmetaTimestampStorage for replay protection

silva-fj committed:
Implement omni_submitUserOpWithAuth RPC method

silva-fj committed:
Wire up wildmeta dependencies in RPC server

silva-fj committed:
Update Cargo.lock

silva-fj committed:
Add wildmeta-api crate

silva-fj committed:
Add wildmeta-api to workspace

silva-fj committed:
Update rpc-server to use wildmeta-api crate

silva-fj committed:
Update executor-worker to use wildmeta-api crate

silva-fj committed:
Fix clippy warning for unused field

silva-fj committed:
fix(rpc): refactor wildmeta signature verification

silva-fj committed:
fix(rpc): fix imports

silva-fj committed:
refactor(rpc): cleanup formatting and error handling

silva-fj committed:
refactor(rpc): rename timestamp verification function

silva-fj committed:
fix(rpc): update EVM address derivation logic

silva-fj committed:
feat(wildmeta-api): implement verify_hyperliquid_link client

silva-fj committed:
feat(config): add OE_WILDMETA_API_URL configuration

silva-fj committed:
feat(executor): use WildmetaApiClient when API URL is configured

silva-fj committed:
chore: update Cargo.lock

silva-fj committed:
add comments

kziemianek committed:
adjust signature to recent changes

kziemianek committed:
update entrypoint simulations bytecode

silva-fj committed:
feat(rpc-server): add mocks for wildmeta and signer-client

Add mock dependencies for `wildmeta-api`, `signer-client`, and
`executor-storage` to dev-dependencies in Cargo.toml. Update tests in
`get_health.rs`, `get_shielding_key.rs`, and `test_protected_method.rs`
to use `MockWildmetaApi` and `WildmetaTimestampStorage`

silva-fj committed:
refactor(auth): change business_json to String for WildmetaHl

Change the `business_json` field in `ClientAuth::WildmetaHl` from
`serde_json::Value` to `String` for improved flexibility. Update
signature verification logic to accept a string and parse as needed.
Add comprehensive tests for signature verification and JSON parsing.

kziemianek committed:
add unit test for init_code parsing

kziemianek committed:
remove userop ownership check

kziemianek committed:
remove redundant variable

kziemianek committed:
Merge branch 'dev' into p-1602-worker-code-to-initiate-the-deposit

kziemianek committed:
remove unused

kziemianek committed:
resolve clippy warnings and taplo fmt

silva-fj committed:
Merge branch 'p-1602-worker-code-to-initiate-the-deposit' into p-1621-stateless-wildmeta-userop-submission

Resolved conflicts by combining both wildmeta API and mailer functionality

silva-fj committed:
fix(rpc-server): restore rpc_clients field to RpcContext

silva-fj committed:
fix(executor-worker): create and pass rpc_clients to RPC server

silva-fj committed:
test(rpc-server): update tests to include rpc_clients parameter

silva-fj committed:
refactor(submit_user_op_with_auth): simplify sender address validation and remove unused code

silva-fj committed:
refactor(rpc-server): remove unused rpc_clients references and simplify code

silva-fj committed:
Merge remote-tracking branch 'origin/dev' into p-1621-stateless-wildmeta-userop-submission

# Conflicts:
#	tee-worker/omni-executor/aa-contracts/docker-deploy.sh
#	tee-worker/omni-executor/config-loader/src/config.rs
#	tee-worker/omni-executor/docker/entrypoint.sh
#	tee-worker/omni-executor/executor-worker/src/main.rs
#	tee-worker/omni-executor/rpc-server/Cargo.toml
#	tee-worker/omni-executor/rpc-server/src/methods/omni/get_health.rs
#	tee-worker/omni-executor/rpc-server/src/methods/omni/get_shielding_key.rs
#	tee-worker/omni-executor/rpc-server/src/methods/omni/mod.rs
#	tee-worker/omni-executor/rpc-server/src/methods/omni/test_protected_method.rs
#	tee-worker/omni-executor/rpc-server/src/server.rs

Kailai-Wang commented on tee-worker/omni-executor/config-loader/src/config.rs:
```diff
@@ -320,6 +325,15 @@ impl ConfigLoader {
+			(
+				"omni_wallet_implementation_address",
+				EnvVar {
+					env_key: "OE_OMNI_WALLET_IMPLEMENTATION_ADDRESS",
```
Is it only used in tests/local-deployment? I'm not sure why we need that

silva-fj replied:
Let me check, this is not mine.. I must have included it by accident when fixing the merging conflicts

Kailai-Wang commented on tee-worker/omni-executor/rpc-server/src/methods/omni/submit_user_op_with_auth.rs:
```diff
@@ -0,0 +1,311 @@
+					main_address,
+					signature,
+					login_type,
+				} => {
```
We still miss one check:
the `main_address` matches the `user_id` => for evm `user_id` we simply verify they are the same, for others (e.g. email) we'll need the help of signer

silva-fj replied:
We do this check, look at lines 114 and 139:
```
if user_address.to_lowercase() != main_address.to_lowercase() {...}

if derived_address.to_lowercase() != main_address.to_lowercase() {...}
```

Kailai-Wang replied:
Ah I missed that, thank you!

Kailai-Wang reviewed (approved):
Looks good in general, I approve it now

kziemianek commented on tee-worker/omni-executor/executor-worker/src/main.rs:
```diff
@@ -512,6 +512,21 @@ async fn main() -> Result<(), ()> {
+						Arc::new(Box::new(wildmeta_api::WildmetaApiClient::new(url.clone())))
+					}
+				} else {
+					Arc::new(Box::new(wildmeta_api::MockWildmetaApi))
```
Could it make any harm due to missconfig ? 
I would suggest enabling it more explicitly.

silva-fj replied:
resolved

silva-fj committed:
refactor: remove wallet implementation address from config

silva-fj committed:
Make wildmeta_api_url required and remove mock usage

silva-fj committed:
feat(rpc): add omni_submitUserOpTest test method

Introduced the omni_submitUserOpTest RPC method for testing user operation
submission. This includes parameter validation, address decoding, and
integration with the native task handler. Also registered the new method in
the omni module.

silva-fj committed:
refactor: minor cleanup and formatting in omni modules
